src/uu/hashsum/src/hashsum.rs

```diff
@@ -135,114 +204,138 @@ fn detect_algo(
+fn handle_sha3_flag(matches: &ArgMatches) -> Option<(&'static str, Box<dyn Digest>, usize)> {
+    if matches.get_flag("sha3") {
+        match matches.get_one::<usize>("bits") {
+            Some(224) => Some((
+                "SHA3-224",
+                Box::new(Sha3_224::new()) as Box<dyn Digest>,
+                224,
+            )),
+            Some(256) => Some((
+                "SHA3-256",
+                Box::new(Sha3_256::new()) as Box<dyn Digest>,
+                256,
+            )),
+            Some(384) => Some((
+                "SHA3-384",
+                Box::new(Sha3_384::new()) as Box<dyn Digest>,
+                384,
+            )),
+            Some(512) => Some((
+                "SHA3-512",
+                Box::new(Sha3_512::new()) as Box<dyn Digest>,
+                512,
+            )),
+            Some(_) => crash!(
+                1,
+                "Invalid output size for SHA3 (expected 224, 256, 384, or 512)"
             ),
+            None => crash!(1, "--bits required for SHA3"),
+        }
+    } else {
+        None
+    }
+}
+
+/// Creates a hasher instance based on the command-line flags.
+///
+/// # Arguments
+///
+/// * `matches` - A reference to the `ArgMatches` object containing the command-line arguments.
+///
+/// # Returns
+///
+/// Returns a tuple containing the algorithm name, the hasher instance, and the output length in bits.
+///
+/// # Panics
+///
+/// Panics if multiple hash algorithms are specified or if a required flag is missing.
+fn create_algo_from_flags(matches: &ArgMatches) -> (&'static str, Box<dyn Digest>, usize) {
+    let mut alg: Option<Box<dyn Digest>> = None;
+    let mut name: &'static str = "";
+    let mut output_bits = 0;
+    let mut set_or_crash = |n, val, bits| {
+        if alg.is_some() {
+            crash!(1, "You cannot combine multiple hash algorithms!");
+        };
+        name = n;
+        alg = Some(val);
+        output_bits = bits;
+    };
+
+    if matches.get_flag("md5") {
+        set_or_crash("MD5", Box::new(Md5::new()), 128);
+    }
+    if matches.get_flag("sha1") {
+        set_or_crash("SHA1", Box::new(Sha1::new()), 160);
+    }
+    if matches.get_flag("sha224") {
+        set_or_crash("SHA224", Box::new(Sha224::new()), 224);
+    }
+    if matches.get_flag("sha256") {
+        set_or_crash("SHA256", Box::new(Sha256::new()), 256);
+    }
+    if matches.get_flag("sha384") {
+        set_or_crash("SHA384", Box::new(Sha384::new()), 384);
+    }
+    if matches.get_flag("sha512") {
+        set_or_crash("SHA512", Box::new(Sha512::new()), 512);
+    }
+    if matches.get_flag("b2sum") {
+        set_or_crash("BLAKE2", Box::new(Blake2b::new()), 512);
+    }
+    if matches.get_flag("b3sum") {
+        set_or_crash("BLAKE3", Box::new(Blake3::new()), 256);
+    }
+    if let Some((n, val, bits)) = handle_sha3_flag(matches) {
+        set_or_crash(n, val, bits);
+    }
```

Comment:
I would handle the "sha3" flag itself in a similar way to how the other flags are handled. This allows you to reuse the `create_sha3` function and makes `handle_sha3_flag` obsolete. Suggestion:

```rust
if matches.get_flag("sha3") {
    let (n, val, bits) = create_sha3(matches);
    set_or_crash(n, val, bits);
}
```